### PR TITLE
Optional Numpy Compute Backend

### PR DESCRIPTION
**Pull request by cremebrule**

### CODE REVIEW

File: omnigibson/robots/active_camera_robot.py [hunk 1]
```diff
@@ -114,7 +115,7 @@ def _default_controller_config(self):
 
         return cfg
 
-    @property
+    @cached_property
```
Comment (cgokmen) [inline omnigibson/robots/active_camera_robot.py:118]: Does this do anything really? Like will it cache the child's implementation of this method?

File: omnigibson/utils/config_utils.py [hunk 2]
```diff
@@ -2,6 +2,7 @@
 import json
 import os
 
+import numpy as np
```
Comment (cgokmen) [inline omnigibson/utils/config_utils.py:5]: Why is this necessary?

File: setup.py [hunk 3]
```diff
@@ -48,6 +48,8 @@
         "graphviz>=0.20",
         "matplotlib>=3.0.0",
         "lxml>=5.3.0",
+        "numba>=0.60.0",
+        "cffi~=1.16.0",
```
Comment (cgokmen) [inline setup.py:52]: Do we need the tilde or can we live with anything greater than this?

**Review by cgokmen — APPROVED** [review]

File: omnigibson/robots/active_camera_robot.py (hunk 1 above)
Comment (cremebrule) [reply omnigibson/robots/active_camera_robot.py:118]: it's effectively a lookup within itself so the parent and child should each have their own cache AFAIK

File: omnigibson/utils/config_utils.py (hunk 2 above)
Comment (cremebrule) [reply omnigibson/utils/config_utils.py:5]: sorry, not needed (was part of WIP) -- removed!

File: setup.py (hunk 3 above)
Comment (cremebrule) [reply setup.py:52]: This definitely conflicted with something else, but I don't quite remember what -- I'm testing now with a newer version and it seems to now work fine, so maybe we can live with a higher version

File: omnigibson/controllers/osc_controller.py [hunk 4]
```diff
@@ -389,31 +387,36 @@ def compute_control(self, goal_dict, control_dict):
...
-            ee_mat=T.quat2mat(ee_quat),
-            ee_vel=ee_vel,
+            ee_pos=cb.as_float32(ee_pos),
+            ee_mat=cb.as_float32(cb.T.quat2mat(ee_quat)),
+            ee_lin_vel=cb.as_float32(ee_vel[:3]),
+            ee_ang_vel_err=cb.as_float32(
+                cb.T.quat2axisangle(
+                    cb.T.quat_multiply(cb.T.axisangle2quat(-ee_vel[3:]), cb.T.axisangle2quat(base_ang_vel))
+                )
+            ),
```
Comment (ChengshuLi) [inline omnigibson/controllers/osc_controller.py:419]: nit: maybe compute this first and then feed in? Also should we use orientation_error here or is this different?
Or is it cleaner to just feed in `ee_ang_vel` and do this computation within the `compute_osc_torques` function?

File: omnigibson/robots/holonomic_base_robot.py [hunk 5]
```diff
@@ -302,23 +307,25 @@ def _postprocess_control(self, control, control_type):
         u_vec, u_type_vec = super()._postprocess_control(control=control, control_type=control_type)
 
         # Change the control from base_footprint_link ("base_footprint") frame to root_link ("base_footprint_x") frame
-        base_orn = self.base_footprint_link.get_position_orientation()[1]
-        root_link_orn = self.root_link.get_position_orientation()[1]
+        base_orn = ControllableObjectViewAPI.get_position_orientation(self.articulation_root_path)[1]
```
Comment (ChengshuLi) [inline omnigibson/robots/holonomic_base_robot.py:310]: I think this is wrong? And Also, why are we calling `ControllableObjectViewAPI` directly?

**Review by ChengshuLi — APPROVED** [review]

File: omnigibson/controllers/null_joint_controller.py [hunk 6]
```diff
@@ -58,7 +57,7 @@ def __init__(
                 applied
         """
         # Store values
-        self._default_command = th.zeros(len(dof_idx)) if default_command is None else default_command
+        self._default_command = cb.zeros(len(dof_idx)) if default_command is None else default_command
```
[the remark this answers is not in the document]
Comment (ChengshuLi) [reply omnigibson/controllers/null_joint_controller.py:60]: TODO: update to default_goal

File: omnigibson/utils/transform_utils_np.py [hunk 7]
```diff
@@ -0,0 +1,1345 @@
...
+    n_elements = mat.shape[0]
+    out = np.zeros(n_elements, dtype=np.float32)
+    for i in prange(n_elements):
+        vec = mat[i]
+        out[i] = np.sqrt(np.sum(vec * vec))
+    return out
+
+
+@jit(nopython=True)
+def mat2quat_batch(rmat):
```
Comment (ChengshuLi) [inline omnigibson/utils/transform_utils_np.py:408]: figure out if this is used

File: omnigibson/utils/transform_utils_np.py [hunk 8]
```diff
@@ -0,0 +1,1345 @@
...
+from scipy.spatial.transform import Rotation as R
+
+PI = np.pi
+EPS = np.finfo(float).eps * 4.0
+
+# axis sequences for Euler angles
+_NEXT_AXIS = [1, 2, 0, 1]
+
+# map axes strings to/from tuples of inner axis, parity, repetition, frame
+_AXES2TUPLE = {
```
Comment (ChengshuLi) [inline omnigibson/utils/transform_utils_np.py:20]: make sure one-to-one correspondence with `transform_utils.py`

File: omnigibson/utils/transform_utils_np.py (hunk 7 above)
Comment (cremebrule) [reply omnigibson/utils/transform_utils_np.py:408]: it is

File: omnigibson/controllers/ik_controller.py [hunk 9]
```diff
@@ -344,16 +332,20 @@ def compute_control(self, goal_dict, control_dict):
 
     def compute_no_op_goal(self, control_dict):
         # No-op is maintaining current pose
+        target_pos = cb.copy(control_dict[f"{self.task_name}_pos_relative"])
+        target_quat = cb.copy(control_dict[f"{self.task_name}_quat_relative"])
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/controllers/ik_controller.py:336]: removed

File: omnigibson/controllers/ik_controller.py [hunk 10]
```diff
@@ -374,9 +366,82 @@ def _compute_no_op_action(self, control_dict):
...
-            target_quat=(4,),
+            target_ori_mat=(3, 3),
         )
 
     @property
     def command_dim(self):
         return IK_MODE_COMMAND_DIMS[self.mode]
+
+
+import torch as th
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/controllers/ik_controller.py:377]: to denote that these functions are only used for those specific functions

File: omnigibson/controllers/ik_controller.py [hunk 11]
```diff
@@ -374,9 +366,82 @@ def _compute_no_op_action(self, control_dict):
...
+    target_joint_pos = q + delta_j
+
+    # Clip values to be within the joint limits
+    return target_joint_pos.clip(q_lower_limit, q_upper_limit)
+
+
+# Set these as part of the backend values
+setattr(_ComputeBackend, "compute_ik_qpos", None)
+setattr(_ComputeTorchBackend, "compute_ik_qpos", _compute_ik_qpos_torch)
+setattr(_ComputeNumpyBackend, "compute_ik_qpos", _compute_ik_qpos_numpy)
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/controllers/ik_controller.py:447]: refactored to make adding custom functions more clean and less hacky -- now explicitly calls public function `add_compute_function()`

File: omnigibson/controllers/multi_finger_gripper_controller.py [hunk 12]
```diff
@@ -117,6 +118,11 @@ def __init__(
             command_output_limits=command_output_limits,
         )
 
+    def _generate_default_command_output_limits(self):
+        command_output_limits = super()._generate_default_command_output_limits()
+
+        return cb.mean(command_output_limits[0]), cb.mean(command_output_limits[1])
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/controllers/multi_finger_gripper_controller.py:124]: done

File: omnigibson/controllers/null_joint_controller.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/controllers/null_joint_controller.py:60]: done

File: omnigibson/controllers/osc_controller.py (hunk 4 above)
Comment (cremebrule) [reply omnigibson/controllers/osc_controller.py:419]: I think the computation is roughly the same. orientation_error is optimized for matrix multiplication, and since the inputs are in axis angle form, we'd have to do multiple transforms (axisangle -> quat -> matrix) since axisangle -> matrix is currently not implemented directly

File: omnigibson/utils/transform_utils_np.py (hunk 8 above)
Comment (cremebrule) [reply omnigibson/utils/transform_utils_np.py:20]: done
